### PR TITLE
[Fix #2321] Don't replace reduce with each_with_object if accumulator var is assigned to

### LINKED ISSUE #2321: each_with_object style must respect immutable method_args
each_with_object does not work on immutable objects like integer.
http://stackoverflow.com/questions/19064209/how-is-each-with-object-supposed-to-work

In case of 
[1, 2, 3].reduce(0) { |a, e| a += e }
rubocop must leave it as is. 

rubocop -V
0.34.2 (using Parser 2.2.2.5, running on ruby 2.2.2 x86_64-linux)

Thanks.

### PR DESCRIPTION
**Pull request by alexdowad**
Otherwise we would change the meaning of the code!

### CODE REVIEW

File: lib/rubocop/cop/style/each_with_object.rb [hunk 1]
```diff
@@ -16,6 +19,7 @@ module Style
       class EachWithObject < Cop
         MSG = 'Use `each_with_object` instead of `%s`.'
         METHODS = [:inject, :reduce]
+        ASSIGNMENT_TYPES = [:lvasgn, :casgn, :cvasgn, :gvasgn, :ivasgn, :masgn]
```
Comment (jonas054) [inline lib/rubocop/cop/style/each_with_object.rb]: This is already available as `EQUALS_ASGN_NODES`, mixed in from `Util`.

Comment (alexdowad) [inline lib/rubocop/cop/style/each_with_object.rb]: Nice catch. I didn't know about this. Updating PR now.

**Comment by alexdowad** [comment]
Updated according to feedback from @jonas054. Thank you!

**Comment by bbatsov** [comment]
We can further optimize this if we simply check whether the initial value is some immutable literal, as the subsequent checks would be redundant then.

**Comment by alexdowad** [comment]
OK. So in that case, we don't even consider changing to `with_each_object`. Makes sense.

**Comment by alexdowad** [comment]
Updated according to @bbatsov's feedback.

**Comment by bbatsov** [comment]
Looks good, but we have the usual conflict in the changelog. :-)

**Comment by alexdowad** [comment]
This constant rebasing is getting tedious.

**Comment by bbatsov** [comment]
True, but now it's a good time for a rebase.

**Comment by bbatsov** [comment]
P.S. I contemplated auto-generating the changelog at some point, but then I'd have to explain every the conventions for the commit messages, so the generated changelog is meaningful. In  the end of the day there are no silver bullets (as usual).

**Comment by alexdowad** [comment]
Rebased.

**Comment by alexdowad** [comment]
If you merge this one and are game for more, #2366 has been rebased onto this one, so it should go through without another rebase. It is quite simple, so I anticipate the review will be fairly quick.

**Comment by alexdowad** [comment]
BTW. Have you ever tried downloading patches with `wget` and applying them with `git am`, rather than using the GitHub "Merge" button? It makes your git history look much nicer (no merge commits), and may help to provide an antidote to this constant rebasing.

If you ff-merge (not adding merge commits), then I could send you a series of PRs, each one based on the previous one. You could merge them in order without any rebasing.

**Comment by alexdowad** [comment]
...Only if you decided to reject a PR, then I would rebase the following ones. Can we try?

**Comment by bbatsov** [comment]
From time to time I manually rebase some branches if the author disappears. I'm well aware of all the advantages of this approach, but unfortunately I'm so busy at work that I favour the approach which is the least work for me. Anyways, I'm fine with applying manually your PRs, as I really value the time you're taking to clean up our huge backlog of issues.

**Comment by alexdowad** [comment]
I'm happy to do whatever is easy for you, but the approach I am suggesting takes very little effort and will reduce the need for all this rebasing. Up to you.

**Pull request merged**